consensus/core/src/dag_state.rs

```diff
@@ -83,11 +83,30 @@ impl DagState {
     /// Accepts a block into DagState and keeps it in memory.
     pub(crate) fn accept_block(&mut self, block: VerifiedBlock) {
         let block_ref = block.reference();
+        // Ensure we don't write multiple unique blocks per slot for our own index
+        if block_ref.author == self.context.own_index {
+            let existing_blocks = self.get_uncommitted_blocks_at_slot(block_ref.into());
```

Comment:
`get_uncommitted_blocks_at_slot()` has a check that the slot is above last committed leader round, which may not be true here even though the block is indeed uncommitted. At least one of the checks needs to be updated. We can
(1) remove this check or move it to Core
(2) have a `get_blocks_at_slot()` function that is "best effect".
(3) remove the round check in `get_uncommitted_blocks_at_slot()`
I prefer option 1 because this seems to be an invariant that best enforced inside proposing logic.

Reply:
Option 1 sounds good to me as well. I find it very convenient this check to be here, but ultimately we want Core to be able to detect this.

Reply:
Ok to leave TODO. I think the best invariant to check for the most recent own block is that its ancestors contain the previous own block with the next-highest round. We can iterate on this.